RFC/rfcBBL209/README.md

```diff
@@ -0,0 +1,104 @@
+   pieces = getChunks(D)
+
+   var S state
+   for i, p := range pieces {
+      S = process(S, p) // Call this result S_i
+   }
+
+   return finalize(S) // Call this H, the final hash
+}
+```
+
+From the above we can see that:
+
+1. At any point in the process of hashing D we could stop, say after piece `j`, save the state `S_j` and then resume later
+2. We can always calculate the final hash `H` given only `S_j` and all the pieces `P_j+1..P_n-1`
+
+The implication for Bitswap is that if each piece size is not more than 1MiB then we can send the file **backwards** in 1MiB increments. In particular a server can send `(S_n-2, P_n-1)` and the client can use that to compute that `P_n-1` is in fact the last part of the data associated with the final hash `H`. The server can then send `(S_n-3, P_n-2)` and the client can calculate that `P_n-2` is the last block of `S_n-2` and therefore also the second to last block of `H`, and so on.
+
+#### Extension
+
+This scheme requires linearly downloading a file which is quite slow with even modest latencies. However, utilizing a scheme like [RFC|BB|L2 - Speed up Bitswap with GraphSync CID only query](https://github.com/protocol/beyond-bitswap/issues/25) (i.e. downloading metadata manifests up front) we can make this fast/parallelizable
+
+#### Security
+
+In order for this scheme to be secure it must be true that there is only a single pair `(S_i-1, P_i)` that can be produced to match with `S_i`. If the pair must be of the form `(S_i-1, P_malicious)` then this is certainly true since otherwise one could create a collision on the overall hash function. However, given that there are two parameters to vary it seems possible this could be computationally easier than finding a collision on the overall hash function.
+
+#### SHA-3
+
+While SHA-3 is not a Merkle–Damgård construction it follows the same psuedocode structure above
+
+### Example: Tree constructions like Blake3, Kangaroo-Twelve, or ParallelHash
+
+In tree constructions we are not restricted to downloading the file backwards and can instead download the parts of the file the we are looking for, which includes downloading the file forwards for sequential streaming.
+
+There is detail about how to do this for Blake3 in the [Blake3 paper](https://github.com/BLAKE3-team/BLAKE3-specs/blob/master/blake3.pdf) section 6.4, Verified Streaming
+
+### Implementation Plan
+
+#### Bitswap changes
+
+* When a server responds to a request for a block if the block is too large then instead send a traversal order list of the block as defined by the particular hash function used (e.g. linear and backwards for SHA-1,2,3)
+* Large Manifests
+  * If the list is more than 1MiB long then only send the first 1MiB along with an indicator that the manifest is not complete
+  * When the client is ready to process more of the manifest then it can send a request WANT_LARGE_BLOCK_MANIFEST containing the multihash of the entire large block and the last hash in the manifest
```

Comment:
How about: Instead of special-casing the manifest file (and having to deal with large vs small manifests), recursively treat the manifest as a downloadable artifact:

If the manifest is small (1<MB), send the whole manifest in the response, otherwise send the manifest of the manifest.